x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/changepoint/SpikeAndDipDetector.java

```diff
@@ -0,0 +1,174 @@
+
+        extent = Math.min(extent, values.length - 1);
+
+        final int argmax = argmax(values, 0, values.length, negate);
+
+        // Find the maximum average interval of width extent which includes argmax.
+        int maxStart = Math.max(0, argmax + 1 - extent);
+        int maxEnd = Math.min(maxStart + extent, values.length);
+        double maxAvg = average(values, maxStart, maxEnd, negate);
+        for (int start = maxStart + 1; start <= argmax; start++) {
+            if (start + extent >= values.length) {
+                break;
+            }
+            double average = average(values, start, start + extent, negate);
+            if (average > maxAvg) {
+                maxStart = start;
+                maxAvg = average;
+            }
+        }
+
+        return new SpikeOrDip(argmax, maxStart, maxStart + extent);
+    }
+
+    private interface ExcludedPredicate {
+        boolean exclude(int index);
+    }
+
+    private double[] removeIf(ExcludedPredicate should, double[] values) {
+        int numKept = 0;
+        for (int i = 0; i < values.length; i++) {
+            if (should.exclude(i) == false) {
+                numKept++;
+            }
+        }
+        double[] newValues = new double[numKept];
+        int j = 0;
+        for (int i = 0; i < values.length; i++) {
+            if (should.exclude(i) == false) {
+                newValues[j++] = values[i];
+            }
+        }
+        return newValues;
+    }
+
+    private final int numValues;
+    private final int dipIndex;
+    private final int spikeIndex;
+    private final double dipValue;
+    private final double spikeValue;
+    private final KDE spikeTestKDE;
+    private final KDE dipTestKDE;
+
+    SpikeAndDipDetector(double[] values) {
+
+        numValues = values.length;
+
+        if (values.length < 4) {
+            dipIndex = -1;
+            spikeIndex = -1;
+            dipValue = Double.NaN;
+            spikeValue = Double.NaN;
+            spikeTestKDE = new KDE(values, 0.8);
+            dipTestKDE = new KDE(values, 0.8);
+            return;
+        }
+
+        // Usually roughly 10% of values.
+        int spikeLength = Math.max((int) (0.1 * values.length + 0.5), 2) - 1;
+
+        SpikeOrDip dip = findSpikeOrDip(values, spikeLength, true);
+        SpikeOrDip spike = findSpikeOrDip(values, spikeLength, false);
+
+        dipIndex = dip.index();
+        spikeIndex = spike.index();
+        dipValue = dip.value(values);
+        spikeValue = spike.value(values);
+
+        double[] dipKDEValues = removeIf((i) -> (dip.includes(i) || i == spike.index()), values);
+        double[] spikeKDEValues = removeIf((i) -> (spike.includes(i) || i == dip.index()), values);
+
+        // We purposely over smooth to only surface visually significant spikes and dips.
+        dipTestKDE = new KDE(dipKDEValues, 1.36);
+        spikeTestKDE = new KDE(spikeKDEValues, 1.36);
```

Comment:
each `KDE` constructor copies and sorts the values. Seems unnecessary. We should sort `dipKDEValues` & `spikeKDEValues` before passing them in and remove the array copy. No need to copy these things twice.

Reply:
(I wanted "move" ;)), but I changed it along these lines to elide copies.